symengine/polys.cpp

```diff
@@ -0,0 +1,368 @@
+            if ((*i)->get_name() == (*j)->get_name()) {
+                present = true;
+            }
+        }
+        if (present == false) {
+            vars_.push_back(*i);
+            c_2++;
+        }
+        present = false;
+    }
+    for (auto &a: p->polys_set_) {
+        std::vector<long long> old(p->vars_.size());
+        m_pair temp;
+
+        ka::decode(old, a.first);
+        for (int i = 0; i < c_2; i++) {
+            old.push_back(0);
+        }
+        temp.first = ka::encode(old);
+        temp.second = a.second;
+        p_new.polys_set_.insert(temp);
+    }
+    for (auto &a: q->polys_set_) {
+        if (p->vars_.empty()) {
+            q_new.polys_set_ = q->polys_set_;
+        } else {
+            std::vector<long long> new_(c_1);
+            std::vector<long long> old(q->vars_.size());
+            m_pair temp;
+            ka::decode(old, a.first);
+            auto i = new_.begin();
+            bool present = false;
+            int j_c = 0, k_c = 0;
+            for (auto j = vars_.begin(); j < vars_.end(); j++) {
+                for (auto k = q->vars_.begin(); k < q->vars_.end(); k++) {
+                    if ((*j)->get_name() == (*k)->get_name()) {
+                        present = true;
+                        new_[j_c] = old[k_c];
+                    }
+                    k_c++;
+                }
+                if (present == false)
+                    new_.insert(i, 0);
+                if (i != new_.end())
+                    i++;
+                present = false;
+                j_c++;
+                k_c = 0;
+            }            
+            temp.first = ka::encode(new_);
+            temp.second = a.second;
+            q_new.polys_set_.insert(temp);
+        }
+    }
+}
+
+void _mul_hashset(const hash_set &A, const hash_set &B, hash_set &C) {
+    m_pair temp;
+    int count = 0;
+    for (auto &a: A) {
+        for (auto &b: B) {
+            temp.first = a.first + b.first;
+            size_t bucket = C._bucket(temp);
+            auto it = C.end();
+            if (!C.empty())
+                it = C._find(temp, bucket);
+            if (it == C.end()) {
+                // Check it the load factor of C is too large.
+                if ((double(C.size()) + 1) / C.bucket_count() > C.max_load_factor()) {
+                    // Increase the size of the table.
+                    C._increase_size();
+                    // Recompute the bucker.
+                    bucket = C._bucket(temp);
+                }
+                temp.second = a.second;
+                temp.second *= b.second;
+                C._unique_insert(temp, bucket);
+                C._update_size(C.size() + 1u);
+            } else {
+                piranha::math::multiply_accumulate(it->second,a.second,b.second);
+            }
+            count++;
+        }
+    }
+}
+
+RCP<const Polynomial> mul_poly(RCP <const Polynomial> p, RCP <const Polynomial> q) {
```

Comment:
No, first of all, RCP should be passed as a reference, i.e. `RCP<const Polynomial> &p`, and second since you are not passing the reference around, the argument should be just `const Polynomial &p`.